File: convert_csv_to_json.py
#!/usr/bin/env python3
"""
Script para convertir preguntas.csv a preguntas.json
Formato compatible con la aplicación Concurso Trainer
"""

import csv
import json
import os
from typing import List, Dict, Any

def limpiar_texto(texto: str) -> str:
    """Limpia el texto eliminando caracteres especiales y citaciones"""
    if not texto:
        return ""
    
    # Eliminar citaciones de OAI y otros patrones
    import re
    
    # Patrón para eliminar citaciones como [oai_citation:...]
    texto = re.sub(r'\[oai_citation:[^\]]*\]', '', texto)
    
    # Patrón para eliminar URLs
    texto = re.sub(r'https?://[^\s]+', '', texto)
    
    # Patrón para eliminar parámetros de URL
    texto = re.sub(r'\?utm_source=[^\s]+', '', texto)
    
    # Limpiar espacios extra y caracteres especiales
    texto = texto.strip()
    texto = re.sub(r'\s+', ' ', texto)  # Múltiples espacios a uno solo
    
    return texto

def procesar_csv(csv_file: str) -> List[Dict[str, Any]]:
    """Procesa el archivo CSV y retorna una lista de preguntas estructuradas"""
    preguntas = []
    
    with open(csv_file, 'r', encoding='utf-8') as file:
        reader = csv.DictReader(file)
        
        for row in reader:
            # Limpiar y validar cada campo
            categoria = limpiar_texto(row.get('Categoria', ''))
            pregunta = limpiar_texto(row.get('Pregunta', ''))
            respuesta_correcta = limpiar_texto(row.get('Respuesta Correcta', ''))
            
            # Obtener respuestas incorrectas
            respuestas_incorrectas = []
            for i in range(1, 4):  # Respuesta Incorrecta 1, 2, 3
                key = f'Respuesta Incorrecta {i}'
                if key in row and row[key]:
                    resp_incorrecta = limpiar_texto(row[key])
                    if resp_incorrecta:
                        respuestas_incorrectas.append(resp_incorrecta)
            
            # Validar que tenemos todos los campos necesarios
            if (categoria and pregunta and respuesta_correcta and 
                len(respuestas_incorrectas) >= 2):
                
                pregunta_estructurada = {
                    "categoria": categoria,
                    "pregunta": pregunta,
                    "respuesta_correcta": respuesta_correcta,
                    "respuestas_incorrectas": respuestas_incorrectas
                }
                
                preguntas.append(pregunta_estructurada)
                print(f"✓ Procesada: {categoria} - {pregunta[:50]}...")
            else:
                print(f"⚠ Omitida: Campos incompletos en categoría '{categoria}'")
    
    return preguntas

def generar_json(preguntas: List[Dict[str, Any]], output_file: str):
    """Genera el archivo JSON con el formato correcto"""
    
    # Crear directorio si no existe
    os.makedirs(os.path.dirname(output_file), exist_ok=True)
    
    # Estructura del JSON
    json_data = {
        "categorias": [],
        "preguntas": preguntas
    }
    
    # Extraer categorías únicas
    categorias_unicas = list(set(pregunta["categoria"] for pregunta in preguntas))
    categorias_unicas.sort()  # Ordenar alfabéticamente
    json_data["categorias"] = categorias_unicas
    
    # Escribir archivo JSON
    with open(output_file, 'w', encoding='utf-8') as file:
        json.dump(json_data, file, ensure_ascii=False, indent=2)
    
    print(f"\n✅ Archivo JSON generado: {output_file}")
    print(f"📊 Total de preguntas: {len(preguntas)}")
    print(f"🏷️  Categorías: {', '.join(categorias_unicas)}")

def main():
    """Función principal"""
    csv_file = "preguntas.csv"
    json_file = "data/preguntas.json"
    
    print("🔄 Convirtiendo CSV a JSON...")
    print(f"📁 Archivo de entrada: {csv_file}")
    print(f"📁 Archivo de salida: {json_file}")
    print("-" * 50)
    
    # Verificar que existe el archivo CSV
    if not os.path.exists(csv_file):
        print(f"❌ Error: No se encontró el archivo {csv_file}")
        return
    
    try:
        # Procesar CSV
        preguntas = procesar_csv(csv_file)
        
        if not preguntas:
            print("❌ Error: No se pudieron procesar preguntas del CSV")
            return
        
        # Generar JSON
        generar_json(preguntas, json_file)
        
        print("\n🎉 Conversión completada exitosamente!")
        print("\n📋 Resumen:")
        print(f"   • Preguntas procesadas: {len(preguntas)}")
        print(f"   • Categorías encontradas: {len(set(p['categoria'] for p in preguntas))}")
        print(f"   • Archivo generado: {json_file}")
        
    except Exception as e:
        print(f"❌ Error durante la conversión: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    main() 

File: convert_csv_to_json.js
#!/usr/bin/env node
/**
 * Script para convertir preguntas.csv a preguntas.json
 * Formato compatible con la aplicación Concurso Trainer
 * 
 * Uso: node convert_csv_to_json.js
 */

const fs = require('fs');
const path = require('path');

function limpiarTexto(texto) {
    if (!texto) return "";
    
    // Eliminar citaciones de OAI y otros patrones
    texto = texto.replace(/\[oai_citation:[^\]]*\]/g, '');
    
    // Eliminar URLs
    texto = texto.replace(/https?:\/\/[^\s]+/g, '');
    
    // Eliminar parámetros de URL
    texto = texto.replace(/\?utm_source=[^\s]+/g, '');
    
    // Limpiar espacios extra y caracteres especiales
    texto = texto.trim();
    texto = texto.replace(/\s+/g, ' '); // Múltiples espacios a uno solo
    
    return texto;
}

function procesarCSV(csvContent) {
    const lineas = csvContent.split('\n');
    const headers = lineas[0].split(',').map(h => h.trim());
    
    console.log('📋 Headers encontrados:', headers);
    
    const preguntas = [];
    
    for (let i = 1; i < lineas.length; i++) {
        const linea = lineas[i].trim();
        if (!linea) continue;
        
        // Parsear CSV considerando comas dentro de campos
        const campos = [];
        let campoActual = '';
        let dentroDeComillas = false;
        
        for (let j = 0; j < linea.length; j++) {
            const char = linea[j];
            
            if (char === '"') {
                dentroDeComillas = !dentroDeComillas;
            } else if (char === ',' && !dentroDeComillas) {
                campos.push(campoActual.trim());
                campoActual = '';
            } else {
                campoActual += char;
            }
        }
        campos.push(campoActual.trim()); // Último campo
        
        if (campos.length >= 5) { // Mínimo 5 campos: Categoria, Pregunta, Respuesta Correcta, 2 incorrectas
            const categoria = limpiarTexto(campos[0]);
            const pregunta = limpiarTexto(campos[1]);
            const respuestaCorrecta = limpiarTexto(campos[2]);
            
            const respuestasIncorrectas = [];
            for (let k = 3; k < Math.min(6, campos.length); k++) {
                const respIncorrecta = limpiarTexto(campos[k]);
                if (respIncorrecta) {
                    respuestasIncorrectas.push(respIncorrecta);
                }
            }
            
            // Validar que tenemos todos los campos necesarios
            if (categoria && pregunta && respuestaCorrecta && respuestasIncorrectas.length >= 2) {
                const preguntaEstructurada = {
                    categoria: categoria,
                    pregunta: pregunta,
                    respuesta_correcta: respuestaCorrecta,
                    respuestas_incorrectas: respuestasIncorrectas
                };
                
                preguntas.push(preguntaEstructurada);
                console.log(`✓ Procesada: ${categoria} - ${pregunta.substring(0, 50)}...`);
            } else {
                console.log(`⚠ Omitida: Campos incompletos en categoría '${categoria}'`);
            }
        }
    }
    
    return preguntas;
}

function generarJSON(preguntas, outputFile) {
    // Crear directorio si no existe
    const outputDir = path.dirname(outputFile);
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }
    
    // Estructura del JSON
    const categoriasUnicas = [...new Set(preguntas.map(p => p.categoria))].sort();
    
    const jsonData = {
        categorias: categoriasUnicas,
        preguntas: preguntas
    };
    
    // Escribir archivo JSON
    fs.writeFileSync(outputFile, JSON.stringify(jsonData, null, 2), 'utf8');
    
    console.log(`\n✅ Archivo JSON generado: ${outputFile}`);
    console.log(`📊 Total de preguntas: ${preguntas.length}`);
    console.log(`🏷️  Categorías: ${categoriasUnicas.join(', ')}`);
}

function main() {
    const csvFile = 'preguntas.csv';
    const jsonFile = 'data/preguntas.json';
    
    console.log('🔄 Convirtiendo CSV a JSON...');
    console.log(`📁 Archivo de entrada: ${csvFile}`);
    console.log(`📁 Archivo de salida: ${jsonFile}`);
    console.log('-'.repeat(50));
    
    // Verificar que existe el archivo CSV
    if (!fs.existsSync(csvFile)) {
        console.log(`❌ Error: No se encontró el archivo ${csvFile}`);
        return;
    }
    
    try {
        // Leer archivo CSV
        const csvContent = fs.readFileSync(csvFile, 'utf8');
        
        // Procesar CSV
        const preguntas = procesarCSV(csvContent);
        
        if (preguntas.length === 0) {
            console.log('❌ Error: No se pudieron procesar preguntas del CSV');
            return;
        }
        
        // Generar JSON
        generarJSON(preguntas, jsonFile);
        
        console.log('\n🎉 Conversión completada exitosamente!');
        console.log('\n📋 Resumen:');
        console.log(`   • Preguntas procesadas: ${preguntas.length}`);
        console.log(`   • Categorías encontradas: ${new Set(preguntas.map(p => p.categoria)).size}`);
        console.log(`   • Archivo generado: ${jsonFile}`);
        
    } catch (error) {
        console.log(`❌ Error durante la conversión: ${error.message}`);
        console.error(error);
    }
}

// Ejecutar si se llama directamente
if (require.main === module) {
    main();
}

module.exports = { procesarCSV, generarJSON, limpiarTexto }; 